exp/controllers/awsmachinepool_controller.go

```diff
@@ -392,6 +425,163 @@ func (r *AWSMachinePoolReconciler) reconcileDelete(machinePoolScope *scope.Machi
 	return nil
 }
 
+func reconcileDeleteAWSMachines(ctx context.Context, mp *expclusterv1.MachinePool, client client.Client) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("machinepool", mp.Name, "namespace", mp.Namespace)
+	awsMachineList, err := getAWSMachines(ctx, mp, client)
+	if err != nil {
+		return err
+	}
+	for i := range awsMachineList.Items {
+		awsMachine := awsMachineList.Items[i]
+		if !awsMachine.DeletionTimestamp.IsZero() {
+			// delete the owner Machine resource for the AWSMachine so that CAPI can clean up gracefully
+			machine, err := util.GetOwnerMachine(ctx, client, awsMachine.ObjectMeta)
+			if err != nil {
+				log.V(2).Info("failed to get owner machine", "awsmachine", awsMachine)
+				continue
+			}
+
+			if err := client.Delete(ctx, machine); err != nil {
+				log.V(2).Info("failed to delete owner machine", "awsmachine", awsMachine)
+			}
+		}
+	}
+	return nil
+}
+
+func getAWSMachines(ctx context.Context, mp *expclusterv1.MachinePool, kubeClient client.Client) (*infrav1.AWSMachineList, error) {
+	awsMachineList := &infrav1.AWSMachineList{}
+	labels := map[string]string{
+		clusterv1.MachinePoolNameLabel: mp.Name,
+		clusterv1.ClusterNameLabel:     mp.Spec.ClusterName,
+	}
+	if err := kubeClient.List(ctx, awsMachineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return nil, err
+	}
+	return awsMachineList, nil
+}
+
+func createAWSMachinesIfNotExists(ctx context.Context, awsMachineList *infrav1.AWSMachineList, mp *expclusterv1.MachinePool, existingASG *expinfrav1.AutoScalingGroup, client client.Client, ec2Svc services.EC2Interface) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("machinepool", mp.Name, "namespace", mp.Namespace)
+	log.V(2).Info("create missing awsmachines")
+
+	providerIDs := make(map[string]infrav1.AWSMachine, len(awsMachineList.Items))
+	for i := range awsMachineList.Items {
+		awsMachine := awsMachineList.Items[i]
+		if *awsMachine.Spec.ProviderID != "" {
+			providerIDs[*awsMachine.Spec.ProviderID] = awsMachine
+		}
+	}
+	ownerRef := metav1.OwnerReference{
+		APIVersion: mp.GroupVersionKind().Version,
+		Kind:       mp.Kind,
+		Name:       mp.Name,
+		UID:        mp.UID,
+	}
+
+	for i := range existingASG.Instances {
+		instanceID := existingASG.Instances[i].ID
+		providerID := fmt.Sprintf("aws:///%s/%s", existingASG.Instances[i].AvailabilityZone, instanceID)
+
+		instanceLogger := log.WithValues("providerID", providerID, "instanceID", instanceID, "asg", existingASG.Name)
+		instanceLogger.V(5).Info("checking if machinepool awsmachine is up to date")
+		if _, exists := providerIDs[providerID]; !exists {
```

Comment:
I would prefer an inverse if here saying `exists -> continue` instead of this huge if. :)